### PR TITLE
Concept - Prompt Engineering

### PR DESCRIPTION
**Pull request by haywoodsloan**
## Summary

Adds the following concept docs:
- _Prompt engineering in .NET_
- _Chain of thought prompting_
- _Zero and few-shot learning_

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/ai/conceptual/chain-of-thought-prompting.md](https://github.com/dotnet/docs/blob/e7145174f742d64ce5b2086937d3ac6e4f90da47/docs/ai/conceptual/chain-of-thought-prompting.md) | [Chain of thought prompting](https://review.learn.microsoft.com/en-us/dotnet/ai/conceptual/chain-of-thought-prompting?branch=pr-en-us-40462) |
| [docs/ai/conceptual/prompt-engineering-dotnet.md](https://github.com/dotnet/docs/blob/e7145174f742d64ce5b2086937d3ac6e4f90da47/docs/ai/conceptual/prompt-engineering-dotnet.md) | ["Prompt Engineering"](https://review.learn.microsoft.com/en-us/dotnet/ai/conceptual/prompt-engineering-dotnet?branch=pr-en-us-40462) |
| [docs/ai/conceptual/zero-shot-learning.md](https://github.com/dotnet/docs/blob/e7145174f742d64ce5b2086937d3ac6e4f90da47/docs/ai/conceptual/zero-shot-learning.md) | [Zero-shot and few-shot learning](https://review.learn.microsoft.com/en-us/dotnet/ai/conceptual/zero-shot-learning?branch=pr-en-us-40462) |

### CODE REVIEW

**Comment by haywoodsloan** [comment]
@catbutler owns these docs, I'm just posting the PR for convenience

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (BillWagner) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:87]: Again, here I'd choose a header that makes it clear you mean examples in your prompt, not examples for the reader to try.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (BillWagner) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:91]: Clarifying not to intersperse the examples worked well.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (BillWagner) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:117]: Are the `//` comments meant to be part of the prompt? As written, they will be.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (BillWagner) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:141]: Is "SK" a common abbreviation? If so, you should use it in parentheses when you first introduce "Semantic Kernel (SK)". If we prefer customers to use the full words, we should use them in this article.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (BillWagner) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:150]: Here, raw string literals will look much cleaner:

```suggestion
```csharp
prompt = """
<message role="system">
    You are helping students with US History homework. Answers should not be about other countries except in 
    relation to the US. Provide some supporting information for answers. Format any summaries using lists.
</message>
<message role="user">Which president had the shortest term?<message/>
<message role="assistant">
    William Henry Harrison died just 31 days after his inauguration as president in 1841,
    making his presidency the shortest in U.S. history.
<message/>
<message role="user">
    Instructions: Summarize US Presidential accomplishments, grouped by category.
    Categories: Domestic Policy, US Economy, Foreign Affairs, Space Exploration, Other.
</message>
""";
```
```

File: docs/ai/conceptual/zero-shot-learning.md
Comment (BillWagner) [inline docs/ai/conceptual/zero-shot-learning.md:20]: grammar nit:

```suggestion
With zero-shot learning, you include prompts but not verbatim completions. You can include completions that only consist of cues. Zero-shot learning relies entirely on the model's existing knowledge to generate responses, reducing the amount of tokens created, which can help you control costs. However, zero-shot learning doesn't add to the model's knowledge.
```

File: docs/ai/conceptual/zero-shot-learning.md
Comment (BillWagner) [inline docs/ai/conceptual/zero-shot-learning.md:29]: Here, this does use string interpolation, but I'd still use raw string literals to improve formatting:

```suggestion
prompt = $"""
Instructions: What is the intent of this request?
If you don't know the intent, don't guess; instead respond with "Unknown".
Choices: SendEmail, SendMessage, CompleteTask, CreateDocument, Unknown.
User Input: {request}
Intent: 
""";
```

File: docs/ai/conceptual/zero-shot-learning.md
Comment (BillWagner) [inline docs/ai/conceptual/zero-shot-learning.md:46]: Similar point:

```suggestion
prompt = $"""
Instructions: What is the intent of this request?
If you don't know the intent, don't guess; instead respond with "Unknown".
Choices: SendEmail, SendMessage, CompleteTask, CreateDocument, Unknown.

User Input: Can you send a very quick approval to the marketing team?
Intent: SendMessage

User Input: Can you send the full update to the marketing team?
Intent: SendEmail

User Input: {request}
Intent: 
""";
```

**Review by BillWagner — COMMENTED** [review]
There's a lot of great information here @haywoodsloan 

I had a lot of comments. Hopefully most are straightforward and easy to address.

**Comment by haywoodsloan** [comment]
> There's a lot of great information here @haywoodsloan
> 
> I had a lot of comments. Hopefully most are straightforward and easy to address.

Thanks! For transparency these changes are from @catbutler, I just posted the PR.

**Review by gewarren — COMMENTED** [review]
Posting my comments so far, but not finished.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:32]: I would love to ditch those "This section..." sentences! They are from the concepts template PDF we got early on. I find them clunky, not clarifying. I'm going to suggest it to Sloan and Jesse.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:42]: Me too! I'll revise once I understand better. The idea was don't interleave them because you might confuse the model. But there's also other material that says things at the top and bottom of a prompt get more attention from the model than things in the middle, so perhaps the idea here is that instructions require more attention than examples?

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:70]: I had a bunch of similar suggestions and revised accordingly.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:117]: thanks, I revised!

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:141]: Thanks - I will spell it out for now - I have no idea what we prefer customers to do yet.

File: docs/ai/conceptual/zero-shot-learning.md
Comment (catbutler) [reply docs/ai/conceptual/zero-shot-learning.md:20]: good call

**Review by catbutler — APPROVED** [review]
I've incorped everyone's feedback and revised to touch on embeddings etc.

**Review by catbutler — APPROVED** [review]
Finished incorping suggestions.

**Comment by catbutler** [comment]
@gewarren could you please review/approve when you have time? I fixed all the build warnings and am waiting for the workflows to finish, but I'll need an approving review to merge. Thanks so much!

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:1]: Drop "in" from the file name, per https://review.learn.microsoft.com/en-us/help/platform/seo-optimize-urls?branch=main#name-your-files. Also dotnet should be one word.

image

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:28]:
```suggestion
That's the basic idea of prompt engineering: you add context to the prompt to help the model produce better completions. You can do this by giving the model [*instructions*](#use-instructions-to-tell-the-model-what-to-do), [*examples*](#examples-show-the-model-what-to-do), and [*cues*](#cues).
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:39]:
```suggestion
## Use instructions to tell the model what to do
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:45]:
```suggestion
- _Directives_ tell the model how to behave, but aren't simple commands&mdash;think character setup for an improv actor: **"You're helping students learn about U.S. history, so talk about the U.S. unless they specifically ask about other countries."**
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:26]:
```suggestion
Looks right, but what if your app is supposed to help U.S. history students? Pedro Lascurain's 45-minute term is the shortest term for any president, but he served Mexico&mdash;the students are probably looking for _"William Henry Harrsion."_ Clearly, the app could be more helpful to its intended users if you gave it some context.
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:48]:
```suggestion
Because they're open-ended, directives are more flexible than imperatives:
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:66]:
```suggestion
Adjust your instruction so the model groups by category, and append supporting content that specifies those categories:
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:127]: I don't think Semantic Kernel and Azure OpenAI should link to other pages here, if anything, I'd expect these to link to the sections below. But then you can add those links off to other pages in the sections themselves.

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:194]:
```suggestion
## Extend the reach of your prompt engineering techniques
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:196]:
```suggestion
You can also increase the power of your prompts with two more advanced prompt engineering techniques that are covered in depth in their own articles.
```

Comment (gewarren) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:196]: Style guide says to avoid using "us" and "we".

File: docs/ai/conceptual/zero-shot-learning.md
Comment (gewarren) [inline docs/ai/conceptual/zero-shot-learning.md:18]:
```suggestion
With zero-shot learning, you include prompts but not verbatim completions. You can include completions that only consist of cues. Zero-shot learning relies entirely on the model's existing knowledge to generate responses, which reduces the number of tokens created and can help you control costs. However, zero-shot learning doesn't add to the model's knowledge.
```

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (gewarren) [inline docs/ai/conceptual/prompt-engineering-in-dot-net.md:198]:
```suggestion
- LLMs have token input limits that constrain the amount of text you can fit in a prompt. You can use *embeddings* and *vector database solutions* to reduce the number of tokens you need to represent a given piece of text.
```

File: docs/ai/conceptual/zero-shot-learning.md
Comment (gewarren) [inline docs/ai/conceptual/zero-shot-learning.md:69]:
```suggestion
- Example-based learning doesn't work well for complex reasoning tasks. However, adding instructions can help address this.
```

File: docs/ai/conceptual/zero-shot-learning.md
Comment (gewarren) [inline docs/ai/conceptual/zero-shot-learning.md:70]:
```suggestion
- Few-shot learning requires creating lengthy prompts. Prompts with large number of tokens can increase computation and latency. This typically means increased costs. There's also a limit to the length of the prompts.
```

**Review by gewarren — APPROVED** [review]

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:1]: Thanks! I can't open that link though...

File: docs/ai/conceptual/prompt-engineering-in-dot-net.md
Comment (catbutler) [reply docs/ai/conceptual/prompt-engineering-in-dot-net.md:127]: Thanks! revised

**Review by catbutler — APPROVED** [review]
LGTM @haywoodsloan

**Comment by catbutler** [comment]
@gewarren are you a maintainer? There are a few checks that won't run on this PR until a maintainer approves them. I'm sorry to bug you about this same PR yet again, but I can't get it moving myself. Thank you!

**Comment by haywoodsloan** [comment]
@catbutler closing a reopening this PR, that might also help kick off the checks again.

**Pull request merged**
